certora/specs/HashConsensus.spec

```diff
@@ -0,0 +1,742 @@
+//  6. addMember() - cannot add an empty address as a member
+// Status: Pass
+// https://prover.certora.com/output/80942/d9d92700466f42b9b284bb7c0664314e/?anonymousKey=dfbcd43aa6f2a327b2fbe32e2ed4b51a06f52f3e
+rule addMemberRevertsCorrectly() {
+    env e; env e2;
+    address userA; uint256 quorum;
+    bool isUserAMember = getIsMember(e,userA);
+
+    addMember@withrevert(e, userA, quorum);                 // call addMember to add user userA
+    bool callReverted = lastReverted;
+
+    assert isUserAMember => callReverted;                   // cannot add existing member, rule 5
+    assert (userA == 0) => callReverted;                    // cannot add empty address, rule 6
+}
+
+
+//  7, addMember() - adding a member does not add/remove any other member
+// Status: Pass
+// https://prover.certora.com/output/80942/c6333d3aad5f47a19a3c91bb05f04e72/?anonymousKey=8ee21a783e8c2bf20dec20e40174f309a5958098
+rule addMemberADoesNotModifyMemberB() {
+    env e; env e2;
+    address userA; uint256 quorum; address userB;
+    bool isUserAMemberBefore = getIsMember(e,userA);
+    bool isUserBMemberBefore = getIsMember(e,userB);
+
+    addMember(e, userA, quorum);                            // call addMember to add user userA
+
+    bool isUserAMemberAfter = getIsMember(e2,userA);
+    bool isUserBMemberAfter = getIsMember(e2,userB);
+
+    require userA != userB;                                 // userA and userB are different
+
+    assert isUserBMemberBefore => isUserBMemberAfter;       // adding userA does not remove userB
+    assert !isUserBMemberBefore => !isUserBMemberAfter;     // adding userA does not add userB
+}
+
+
+//  8. addMember() - adding a member increases the total members by 1
+// Status: Pass
+// https://prover.certora.com/output/80942/f7433cb747754da8a635632b6ae01a2d/?anonymousKey=47221c507f526d533d83e2347625027dfb485b9b
+rule addMemberCorrectness() {
+    env e; env e2;
+
+    uint256 lengthOf_memberAddressesBefore; uint256 lengthOf_memberStatesBefore;
+    lengthOf_memberAddressesBefore, lengthOf_memberStatesBefore = getLengthOfArrays(e);  // using helper
+
+    require lengthOf_memberAddressesBefore == lengthOf_memberStatesBefore;  // those arrays represent the same members
+    require lengthOf_memberAddressesBefore < max_uint256;                   // safe assumption
+
+    address userA; uint256 quorum;
+    bool isUserAMemberBefore = getIsMember(e,userA);
+
+    addMember(e, userA, quorum);                            // call addMember to add user userA
+
+    uint256 lengthOf_memberAddressesAfter; uint256 lengthOf_memberStatesAfter;
+    lengthOf_memberAddressesAfter, lengthOf_memberStatesAfter = getLengthOfArrays(e);  // using helper
+    bool isUserAMemberAfter = getIsMember(e2,userA);
+
+    assert !isUserAMemberBefore;                            // cannot add userA if he was a member before
+    assert isUserAMemberAfter;                              // adding userA correctly
```

Comment:
you call addMember(), so it implies that userA wasn't a member and will be a member after. Does it make sense to check it?